src/map/mob.cpp

```diff
@@ -1119,17 +1120,31 @@ int mob_spawn (struct mob_data *md)
 		md->bl.x = md->spawn->x;
 		md->bl.y = md->spawn->y;
 
+		// Search can be skipped if spawn location is fixed
 		if( (md->bl.x == 0 && md->bl.y == 0) || md->spawn->xs || md->spawn->ys )
-		{	//Monster can be spawned on an area.
-			if( !map_search_freecell(&md->bl, -1, &md->bl.x, &md->bl.y, md->spawn->xs, md->spawn->ys, battle_config.no_spawn_on_player?4:0) )
-			{ // retry again later
-				if( md->spawn_timer != INVALID_TIMER )
-					delete_timer(md->spawn_timer, mob_delayspawn);
-				md->spawn_timer = add_timer(tick + battle_config.mob_respawn_time,mob_delayspawn,md->bl.id,0);
-				return 1;
+		{
+			// Officially the area is split into 4 squares, 4 lines and 1 dot and for each of those there is one attempt
+			// We simplify this to trying 8 times in the whole area and then at the center cell even though that's not fully accurate
+			// The search for a spawn location is called twice, meaning the area is doubled in size but with a strong bias towards the center
+			// The center cell has a 1/(xs*ys) chance to be picked, otherwise we call the search function to get the start coordinates
+			if ((md->spawn->xs > 1 || md->spawn->ys > 1) && !rnd_chance(1, md->spawn->xs * md->spawn->ys))
+				map_search_freecell(&md->bl, -1, &md->bl.x, &md->bl.y, md->spawn->xs-1, md->spawn->ys-1, 0, 8);
```

Comment:
Somehow here the logic is missing, if this succeeds.

Also why is this one not using `battle_config.no_spawn_on_player?4:0`?

Reply:
We don't need the logic here because we don't care of it succeeds or not, if it fails x and y remain the same, if it succeeds x and y are already updated (it's call-by-reference).

It's also important not to consider "no_spawn_on_player" here because here we just calculate the "bridge cell" from which we create the real spawn area below (where we then consider the setting).

Reply:
I added another comment already, is this sufficiently clear?